### DISCUSSION TITLE
MCMC for lens_model_DESIJ0215-2909

### DISCUSSION THREAD

**Discussion by numjahan** [AstroBridge/BDLensing#216] [Q&A] [answered]
I am currently working on MCMC for my system. I am constantly facing this error - `WARNING:root:Too few points to create valid contours` even though it is converging in the trace plot. Here, I used `n_run = 10000`. 

Here is my [notebook](https://colab.research.google.com/drive/1BmrQNba2AxKfU1fIrtjSiDAZxy_xaRju?usp=sharing) for MCMC run.
In this notebook, `burnin = 5000` &  `thin = 1`. I already tried using various combinations of burn-in and thinning values, but the error persisted.

What steps can I take to fix this problem? Any advice for resolving the issue and improving the model would be greatly appreciated.

**Comment by ajshajib** [comment]
Is there any difference if you change your current lines

```
samples = samples_mcmc[:, i]
reshaped_samples = samples.reshape((n_walkers, n_step), order="F")
chain[:, :, i] = reshaped_samples
```

with 

```
samples = samples_mcmc[:, i].T
chain[:, :, i] = samples.reshape((n_step, n_walkers)).T
```
?

**Reply by numjahan** [reply]
No, it does not make any difference.

**Comment by ajshajib** [comment, marked as answer]
Try sigma-clipping your chain like this:
```
        samples_mcmc_cut = chain[:, burnin::thin, :].reshape((-1, n_params))

        # sigma-clipping
        for i in range(n_params):
          median = np.median(samples_mcmc_cut[:, i])
          mad = np.median(np.absolute(samples_mcmc_cut[:, i] - median))

          clip_sigma = 4.
          indices = np.argwhere((samples_mcmc_cut[:, i] < (median + clip_sigma*mad/0.6744)) &
                                (samples_mcmc_cut[:, i] > (median - clip_sigma*mad/0.6744)))
          
          samples_mcmc_cut = samples_mcmc_cut[indices[:, 0], :]

        if not samples_mcmc_cut == []:
            n, num_param = np.shape(samples_mcmc_cut)
            print("Shape of samples_mcmc_cut:", samples_mcmc_cut.shape)
            plot = corner.corner(samples_mcmc_cut, labels=param_mcmc, show_titles=True)
```

**Reply by numjahan** [reply]
It is displaying an error - 
![image](https://github.com/ajshajib/BDLensing/assets/92924875/ee82908e-bc91-456b-9263-5326f5710ffb)

**Reply by ajshajib** [reply]
Fixed a bug in the above code, try it now.

**Reply by numjahan** [reply]
Thank you! The issue with the warning message `WARNING:root:Too few points to create valid contours` has been resolved. Here is the resulting corner plot.
![image](https://github.com/ajshajib/BDLensing/assets/92924875/c47f5c86-898d-48ef-9729-ba778bc95a05)
